### DISCUSSION TITLE
OAuth2 with discord, cookie state mismatch

### DISCUSSION THREAD

**Discussion by 10114MGT** [pocketbase/pocketbase#3111] [Q&A]
Hello! I am creating a svelte login page with Pocketbase and I am unable to auth with OAuth2. I created a Login form with authWithPassword and got it working perfectly.  I think the issue is in my `hooks.server.ts` file. 

The error I am getting is in my `routes/callback/+server.ts` file. 
The error: `state does not match expected [expectedState] [state]`
I think the cookie settings are returning mismatched expectedState compared to the state. I think I have tried most of the settings for `pb.authStore.exportToCookie({})`

I am using discord OAuth2.0 and the client ID and secret are in the PocketBase DB
Client is running on `http://localhost:5173`
Pocketbase is running on `http://localhost:8080`

Any help would be greatly appreciated! Thank you!

```typescript
`routes/callback/+server.ts`

import { redirect } from '@sveltejs/kit';
import type { RequestEvent, RequestHandler } from './$types';

export const GET: RequestHandler = async ({ locals, url, cookies }: RequestEvent) => {
	const redirectURL = `${url.origin}/callback`;
	const expectedState = cookies.get('state');
	const expectedVerifier = cookies.get('verifier');

	const state = url.searchParams.get('state');
	const code = url.searchParams.get('code');

	const authMethods = await locals.pb?.collection('users').listAuthMethods();

	if (!authMethods?.authProviders) {
		console.log('auth providers');
		throw redirect(303, '/login');
	}
	const provider = authMethods.authProviders[0];
	if (!provider) {
		console.log('Provider not found');
		throw redirect(303, '/login');
	}

	if (expectedState !== state) {
		console.log('state does not match expected', expectedState, state);
		throw redirect(303, '/login');
	}

	try {
		const loginRegister: any = await locals.pb
			?.collection('users')
			.authWithOAuth2Code(provider.name, code || '', expectedVerifier || '', redirectURL, {
				username: '',
				name: 'Test'
			});

		const { record, meta } = loginRegister;

		await locals.pb?.collection('users').update(record.id, {
			username: meta.username,
			discord_id: meta.id
		});

		console.log(locals.pb?.authStore.model);
	} catch (err) {
		console.log('Error logging in with 0Auth user', err);
	}

	throw redirect(303, '/');
};
```

```typescript
`hooks.server.ts`

import { pb } from '$lib/pocketbase';
import type { Handle } from '@sveltejs/kit';

export const handle: Handle = async ({ event, resolve }) => {
	pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');

	if (pb.authStore.isValid) {
		try {
			await pb.collection('users').authRefresh();
		} catch (_) {
			event.locals.pb.authStore.clear();
		}
	}

	event.locals.pb = pb;
	event.locals.user = structuredClone(pb.authStore.model);

	const response = await resolve(event);

	response.headers.set(
		'set-cookie',
		event.locals.pb.authStore.exportToCookie({ httpOnly: false, sameSite: 'lax', secure: false })
	);
	return response;
};
```

**Comment by ganigeorgiev** [comment]
This is a very frequent question (_see https://github.com/pocketbase/pocketbase/discussions/2498, https://github.com/pocketbase/pocketbase/discussions/2772, https://github.com/pocketbase/pocketbase/discussions/2159 and probably others_) and I'm not really sure how to explain it better (_except maybe create a dedicated examples for each meta framework, but this can get quickly out of hand as there is more than one way of integrating..._).

Where do you set the `state` and `verifier` cookie from the previous step (aka. the "Login with ..." page)?

And why do you need to call `listAuthMethods()` in the redirect handling?

All submitted SvelteKit OAuth2 questions so far seems to use a similar template so I guess all are copy-pasted from somewhere but please try to understand what the code is doing, otherwise you are risking compromising your security.

If you are going to export the cookie with `httpOnly: false`, it might be easier to use the single "all-in-one" call, aka. authenticate in the browser and send back the auth cookie to the server. For this you'll need a SDK instance that runs in the browser. For example:

```js
// browserclient.js
import PocketBase from "pocketbase";

const pb = new PocketBase(url);

pb.authStore.loadFromCookie(document.cookie)

pb.authStore.onChange(() => {
  document.cookie = pb.authStore.exportToCookie({ httpOnly: false });
});

export default pb; // it is safe to be global in this case
```
Then in your login page you can import the above client:

```js
import pb from "./browserclient.js"

...

const authData = await pb.collection("users").authWithOAuth2({ provider: 'discord' });
```


But if you want to interact with PocketBase **only** server-side, then you have to use the "Manual code exchange" flow, as in your code sample. To get a general idea of the expected flow, please check the example in ["Manual code exchange" tab](https://pocketbase.io/docs/authentication/#oauth2-integration).
While the code sample uses plain JS in the browser, the idea is the same - aka. in the second redirect step you need to read from somewhere the **previously initiated**: provider name, state, and verifier (the example uses localStorage, in your case you can use cookie).
